Create rule S6802: Using lambda expressions in loops should be avoided in Blazor markup section (#3215)
SonarSource/rspec: rules/S6802/csharp/rule.adoc

== Why is this an issue?

> In Blazor, recreating https://learn.microsoft.com/en-us/aspnet/core/blazor/components/event-handling#lambda-expressions[lambda expression delegates] for elements or components in a loop can lead to performance issues, especially when rendering a large number of elements.

Reviewer: I would rephrase this to mention Lambda expressions since this is the language feature used. The fact that it creates delegates is an implementation detail.

```suggestion
In Blazor, using https://learn.microsoft.com/en-us/aspnet/core/blazor/components/event-handling#lambda-expressions[lambda expressions] in UI elements in a loop can lead to performance issues, especially when rendering a large number of elements.
```

> For instance, assigning a delegate to the `@onclick` event of each button in a set can slow down rendering speed and negatively impact user experience and performance as Blazor needs to rebuild all the delegates each time the buttons are rendered.

Reviewer: This is a bit confusing since also in the compliant example the fix is to assign a delegate: `@onclick="button.Action"`

[source,csharp,diff-id=1,diff-type=compliant]
----
@foreach (var button in Buttons)
{
    <button @key="button.Id" @onclick="button.Action">  @* Compliant *@
        Button #@button.Id
    </button>
}

@code {
    private List<Button> Buttons { get; set; } = new();

    protected override void OnInitialized()
    {
        for (var i = 0; i < 100; i++)
        {
            var button = new Button();

            button.Action = (e) => DoAction(e, button);

            Buttons.Add(button);
        }
    }

    private void DoAction(MouseEventArgs e, Button button)
    {
        // Do something here
    }

    private class Button
    {
        public string? Id { get; } = Guid.NewGuid().ToString();
        public Action<MouseEventArgs> Action { get; set; } = e => { };
    }
}
----

== Resources

=== Documentation

* Microsoft Learn - https://learn.microsoft.com/en-us/aspnet/core/blazor/performance#avoid-recreating-delegates-for-many-repeated-elements-or-components[ASP.NET Core Blazor performance best practices]
* Microsoft Learn - https://learn.microsoft.com/en-us/aspnet/core/blazor/components/event-handling#lambda-expressions[ASP.NET Core Blazor event handling]

=== Benchmarks

> The results were generated by running the following snippet with https://github.com/dotnet/BenchmarkDotNet[BenchmarkDotNet] and https://github.com/egil/Benchmark.Blazor/tree/main[Benchmark.Blazor]:

Reviewer: We need to remove the snippet mention.

== Why is this an issue?

> In Blazor, using https://learn.microsoft.com/en-us/aspnet/core/blazor/components/event-handling#lambda-expressions[lambda expressions] as https://learn.microsoft.com/en-us/aspnet/core/blazor/components/event-handling?view=aspnetcore-7.0#lambda-expressions[event handlers] when the UI elements are rendered in a loop can lead to negative user experiences and performance issues. This is particularly noticeable when rendering a large number of elements.

Reviewer 2: Missed that one 😅
